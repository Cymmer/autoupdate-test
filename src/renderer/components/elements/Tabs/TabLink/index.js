import React from 'react';
import { Link } from 'react-router-dom';
import cn from 'classnames';
import GLOBALS from 'codechum-app-globals';
import PropTypes from 'prop-types';

import { Icon, Tooltip } from 'components/elements';
import { STUDENT_ROUTES } from 'screen-wrappers/Student/constants';
import styles from '../tabs.module.scss';

import IconButton from '../../Button/IconButton';
import iconButtonTypes from '../../Button/constants/iconButtonTypes';
import tabTypes from '../constants/tabTypes';

const handleClickIfLocked = (e) => {
  e.preventDefault();
};

const TabLink = ({
  className,
  children,
  type,
  active,
  to,
  closeAction,
  id,
  hasNewData,
  isLocked,
}) => {
  const link = (
    <Link
      className={cn(className, {
        [styles[`Tab___${type}___active`]]: active,
        [styles[`Tab___${type}`]]: !active,
        [styles[`Tab___${type}___hasNewData`]]: hasNewData,
        [styles.Tab___withClose]: closeAction && active,
        [styles.Tab___locked]: isLocked,
      })}
      data-test="tabLink"
      id={id}
      to={to}
      onClick={isLocked ? handleClickIfLocked : null}
    >
      {children}
      {closeAction && active && (
        <IconButton
          className={styles.Tab_close}
          data-test="closeButton"
          icon="close"
          type={iconButtonTypes.SOLID.XS}
          onClick={closeAction}
        />
      )}
    </Link>
  );

  return !isLocked ? (
    link
  ) : (
    <span className={styles.Tab_linkWrapper}>
      <Tooltip content={GLOBALS.MESSAGES.PREMIUM_FEATURE_LOCK_MESSAGE}>
        <a href={STUDENT_ROUTES.SHOP} rel="noreferrer" target="_blank">
          <Icon className={styles.Tab_lockedIcon} icon="lock" />
        </a>
      </Tooltip>
      {link}
    </span>
  );
};

TabLink.defaultProps = {
  className: null,
  active: false,
  type: tabTypes.HORIZONTAL.LG,
  to: '#',
  closeAction: null,
  id: null,
  hasNewData: false,
  isLocked: false,
};

TabLink.propTypes = {
  className: PropTypes.string,
  active: PropTypes.bool,
  to: PropTypes.string,
  children: PropTypes.node.isRequired,
  type: PropTypes.oneOf([
    tabTypes.HORIZONTAL.LG,
    tabTypes.HORIZONTAL.SM,
    tabTypes.VERTICAL.LG,
    tabTypes.VERTICAL.SM,
  ]),
  closeAction: PropTypes.func,
  id: PropTypes.string,
  hasNewData: PropTypes.bool,
  isLocked: PropTypes.bool,
};

export default TabLink;
