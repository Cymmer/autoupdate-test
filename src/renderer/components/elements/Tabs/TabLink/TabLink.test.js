import React from 'react';
import { shallow } from 'enzyme';
import { findByTestAttr } from 'codechum-app-utils';
import TabLink from './index';

describe('RENDERING - Tabs > TabLink', () => {
  it('should render with default props', () => {
    const component = shallow(<TabLink>Test</TabLink>);
    expect(component).toMatchSnapshot();
  });

  it('should render the close button if there is a close action and is active', () => {
    const component = shallow(
      <TabLink active closeAction={() => {}}>
        Test
      </TabLink>
    );
    expect(component).toMatchSnapshot();
  });

  it('should render the locked state if locked', () => {
    const component = shallow(
      <TabLink active isLocked closeAction={() => {}}>
        Test
      </TabLink>
    );
    expect(component).toMatchSnapshot();
  });

  it('should render the unlocked state if locked', () => {
    const component = shallow(
      <TabLink active closeAction={() => {}} isLocked={false}>
        Test
      </TabLink>
    );
    expect(component).toMatchSnapshot();
  });
});

describe('EVENTS - Tabs > TabButton', () => {
  it('should have a clickable Link with the correct url', () => {
    const url = 'www.codechum.com';
    const component = shallow(<TabLink to={url}>Test</TabLink>);

    // in testing if the link is clickable, we just check if
    // the correct url was passed to it since this is a 3rd
    // party component
    expect(findByTestAttr(component, 'tabLink').prop('to')).toBe(url);
  });

  it('should have a clickable close button if it has a close button', () => {
    const clickFn = jest.fn();
    const component = shallow(
      <TabLink active closeAction={clickFn}>
        Test
      </TabLink>
    );

    findByTestAttr(component, 'closeButton').simulate('click');

    expect(clickFn).toHaveBeenCalled();
  });
});
