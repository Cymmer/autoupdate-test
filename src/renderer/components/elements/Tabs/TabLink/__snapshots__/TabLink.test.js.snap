// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RENDERING - Tabs > TabLink should render the close button if there is a close action and is active 1`] = `
<Link
  className="Tab___horizontal___lg___active Tab___withClose"
  data-test="tabLink"
  id={null}
  onClick={null}
  to="#"
>
  Test
  <IconButton
    className="Tab_close"
    colorName="blue"
    data-test="closeButton"
    disabled={false}
    icon="close"
    iconClassName={null}
    id={null}
    isLocked={false}
    kind="button"
    onClick={[Function]}
    style={null}
    tabIndex={0}
    type="solid___xs"
  />
</Link>
`;

exports[`RENDERING - Tabs > TabLink should render the locked state if locked 1`] = `
<span
  className="Tab_linkWrapper"
>
  <Tooltip
    className={null}
    content="This is a premium feature. Click the lock icon to learn more"
    placement="bottom"
  >
    <a
      rel="noreferrer"
      target="_blank"
    >
      <Icon
        className="Tab_lockedIcon"
        icon="lock"
        id={null}
        isPulsating={false}
        style={null}
      />
    </a>
  </Tooltip>
  <Link
    className="Tab___horizontal___lg___active Tab___withClose Tab___locked"
    data-test="tabLink"
    id={null}
    onClick={[Function]}
    to="#"
  >
    Test
    <IconButton
      className="Tab_close"
      colorName="blue"
      data-test="closeButton"
      disabled={false}
      icon="close"
      iconClassName={null}
      id={null}
      isLocked={false}
      kind="button"
      onClick={[Function]}
      style={null}
      tabIndex={0}
      type="solid___xs"
    />
  </Link>
</span>
`;

exports[`RENDERING - Tabs > TabLink should render the unlocked state if locked 1`] = `
<Link
  className="Tab___horizontal___lg___active Tab___withClose"
  data-test="tabLink"
  id={null}
  onClick={null}
  to="#"
>
  Test
  <IconButton
    className="Tab_close"
    colorName="blue"
    data-test="closeButton"
    disabled={false}
    icon="close"
    iconClassName={null}
    id={null}
    isLocked={false}
    kind="button"
    onClick={[Function]}
    style={null}
    tabIndex={0}
    type="solid___xs"
  />
</Link>
`;

exports[`RENDERING - Tabs > TabLink should render with default props 1`] = `
<Link
  className="Tab___horizontal___lg"
  data-test="tabLink"
  id={null}
  onClick={null}
  to="#"
>
  Test
</Link>
`;
